File: py/p3.py
#!/usr/bin/env python
#
# Finds largest prime factor of the number 600851475143
#
# (c) 2013 Matthew Rheaume

def solve():
	max_prime = 1
	num = 600851475143

	while (num != 1):
		for i in xrange(2,num + 1):
			if (num % i != 0):
				continue

			num = num / i
			if (i > max_prime):
				max_prime = i
			break

	return max_prime

if __name__ == "__main__":
	print solve()

File: hs/p3.hs
-- Finds the largest prime factor of the number 600851475143
--
-- (c) 2013 Matthew Rheaume

module Main where

largestPrimeFactor :: Int -> Int
largestPrimeFactor 1 = 1
largestPrimeFactor n = max factor (largestPrimeFactor $ n `div` factor) where
	factor = factorize 2 n
	factorize m n
		| n `mod` m == 0 = m
		| otherwise = factorize (m + 1) n

solve :: Int
solve = largestPrimeFactor 600851475143

main :: IO ()
main = putStrLn $ show $ solve
